src/main/java/io/jenkins/plugins/artifact_manager_jclouds/JCloudsArtifactManager.java

```diff
@@ -114,42 +109,124 @@ private String getBlobPath(String key, String path) {
-            Blob blob = blobStore.blobBuilder(blobPath).build();
-            blob.getMetadata().setContainer(provider.getContainer());
-            artifactUrls.put(entry.getValue(), provider.toExternalURL(blob, HttpMethod.PUT));
+        Map<String, Long> fileSizes = workspace.act(new RequestFileSizes(artifacts.values()));
+
+        long multipartSize = S3BlobStoreConfig.get().getMultipartSize();
+        Map<String, BlobStoreProvider.MultipartUploader> multipartUploads = new HashMap<>();
+        try {
+            // Map artifacts to urls for upload
+            List<UploadPartTask> uploadTasks = new ArrayList<>();
+            for (Map.Entry<String, String> entry : artifacts.entrySet()) {
+                String path = "artifacts/" + entry.getKey();
+                String blobPath = getBlobPath(path);
+                Blob blob = blobStore.blobBuilder(blobPath).build();
+                blob.getMetadata().setContainer(provider.getContainer());
+                long fileSize = fileSizes.get(entry.getValue());
+                if (fileSize > multipartSize) {
+                    BlobStoreProvider.MultipartUploader uploader = provider.initiateMultipartUpload(blob);
+
+                    long parts = (fileSize + multipartSize - 1) / multipartSize;
+                    for (int part = 0; part < parts; part++) {
+                        long offset = part * multipartSize;
+                        long limit = Math.min(multipartSize, fileSize - offset);
+
+                        URL url = uploader.toExternalURL(part + 1);
+                        uploadTasks.add(new UploadPartTask(url, entry.getValue(), offset, limit, part + 1));
+                    }
+                    multipartUploads.put(entry.getValue(), uploader);
+                } else {
+                    URL url = provider.toExternalURL(blob, HttpMethod.PUT);
+                    uploadTasks.add(new UploadPartTask(url, entry.getValue(), 0, fileSize, 0));
+                }
+            }
+
+            Map<String, List<BlobStoreProvider.Part>> uploadResult = workspace.act(new UploadPartToBlobStorage(uploadTasks, listener));
+            for (Map.Entry<String, List<BlobStoreProvider.Part>> entry : uploadResult.entrySet()) {
+                BlobStoreProvider.MultipartUploader uploader = multipartUploads.remove(entry.getKey());
+                if (uploader == null) {
+                    continue;
+                }
+                uploader.complete(entry.getValue());
+            }
+            listener.getLogger().printf("Uploaded %s artifact(s) to %s%n", uploadResult.size(), provider.toURI(provider.getContainer(), getBlobPath("artifacts/")));
+        } finally {
+            for (BlobStoreProvider.MultipartUploader uploader : multipartUploads.values()) {
+                try {
+                    uploader.close();
+                } catch (Exception e) {
+                    listener.getLogger().printf("Can't abort multipart upload: %s", e.getMessage());
```

Comment:
Usually would use `Functions.printStackTrace` or similar.

Reply:
I don't know, how to use `Functions.printStackTrace`.
This error is not fatal and usually can be silently ignored. As I known, uncompleted multipart uploads should be removed by S3 bucket lifecycle policy.